src/libs/ReportUtils.js

```diff
@@ -768,22 +779,22 @@ function canDeleteReportAction(reportAction, reportID) {
         return false;
     }
     const isActionOwner = reportAction.actorAccountID === currentUserAccountID;
-    if (isActionOwner && ReportActionsUtils.isMoneyRequestAction(reportAction) && !isSettled(reportAction.originalMessage.IOUReportID)) {
+    const report = getReport(reportID);
```

Comment:
This function becomes a mess...

Do I understand correctly that it's equivalent to...

```js
const report = getReport(reportID);

const isMoneyRequestActionProtectedFromBeingUserDeleted = () => {
    if (ReportActionsUtils.isMoneyRequestAction(reportAction)) {
        // For now, users cannot delete split actions
        const isSplitAction = lodashGet(reportAction, 'originalMessage.type') === CONST.IOU.REPORT_ACTION_TYPE.SPLIT;
        return isSplitAction || isSettled(reportAction.originalMessage.IOUReportID) || isReportApproved(report);
    } else {
        return false;
    }
}

const isActionProtectedFromBeingUserDeleted = () => reportAction.actionName !== CONST.REPORT.ACTIONS.TYPE.ADDCOMMENT ||
        reportAction.pendingAction === CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE ||
        ReportActionsUtils.isCreatedTaskReportAction(reportAction) ||
        reportAction.actorAccountID === CONST.ACCOUNT_ID.CONCIERGE ||
        isMoneyRequestActionProtectedFromBeingUserDeleted();

const isAdmin = () => {
    const policy = lodashGet(allPolicies, `${ONYXKEYS.COLLECTION.POLICY}${report.policyID}`) || {};
    return policy.role === CONST.POLICY.ROLE.ADMIN && !isDM(report);
}

if ((isActionOwner && !isActionProtectedFromBeingUserDeleted()) || isAdmin()) {
    return true;
}

```

Reply:
I think below 2 points look incorrect to me

1) Currently, We are returning false for these 4 conditions. so in your implementation, we need to reverse these conditions
```
        reportAction.actionName !== CONST.REPORT.ACTIONS.TYPE.ADDCOMMENT ||
        reportAction.pendingAction === CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE ||
        ReportActionsUtils.isCreatedTaskReportAction(reportAction) ||
        reportAction.actorAccountID === CONST.ACCOUNT_ID.CONCIERGE ||
```

2) As we are returning early based on conditions, we should avoid checking the `isAdmin` otherwise for ADDCOMMENT also the admin will have the option to delete someone else's comment `|| isAdmin()`. 


So I think combining all parts here won't work as expected.
`(isActionOwner && !isActionProtectedFromBeingUserDeleted()) || isAdmin();`

Reply:
You're right, early returns matter a lot here.

Still, we introduce code duplication and add extra complexity to already complex function.

How about this?

```js
function canDeleteReportAction(reportAction, reportID) {
    const report = getReport(reportID);
    
    const isActionOwner = reportAction.actorAccountID === currentUserAccountID;

    if (ReportActionsUtils.isMoneyRequestAction(reportAction)) {
        // For now, users cannot delete split actions
        const isSplitAction = lodashGet(reportAction, 'originalMessage.type') === CONST.IOU.REPORT_ACTION_TYPE.SPLIT;

        if (isSplitAction || isSettled(reportAction.originalMessage.IOUReportID) || isReportApproved(report)) {
            return false;
        }

        if (isActionOwner) {
            return true;
        }
    }

    if (
        reportAction.actionName !== CONST.REPORT.ACTIONS.TYPE.ADDCOMMENT ||
        reportAction.pendingAction === CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE ||
        ReportActionsUtils.isCreatedTaskReportAction(reportAction) ||
        reportAction.actorAccountID === CONST.ACCOUNT_ID.CONCIERGE
    ) {
        return false;
    }

    const policy = lodashGet(allPolicies, `${ONYXKEYS.COLLECTION.POLICY}${report.policyID}`) || {};
    const isAdmin = policy.role === CONST.POLICY.ROLE.ADMIN && !isDM(report);

    return isActionOwner || isAdmin;
}
```

I proved that these two functions have equivalent truth tables:

```js
function canDeleteAction1(args) {
    if (args.isMoneyRequestAction && args.isSplitAction) {
        return false;
    }

    if (args.isActionOwner && args.isMoneyRequestAction && !args.isActionSettled && !args.isApproved) {
        return true;
    }
    if (
        args.isNotAddComment ||
        args.isPendingDelete ||
        args.isCreatedTask ||
        (args.isMoneyRequestAction && (args.isActionSettled || args.isApproved)) ||
        args.isConcierge
    ) {
        return false;
    }
    if (args.isActionOwner) {
        return true;
    }
    return args.isAdmin;
}

function canDeleteAction2(args) {
    if (args.isMoneyRequestAction) {
        if (args.isSplitAction || args.isActionSettled || args.isApproved) {
            return false;
        } else if (args.isActionOwner) {
            return true;
        }
    }

    if (
        args.isNotAddComment ||
        args.isPendingDelete ||
        args.isCreatedTask ||
        args.isConcierge
    ) {
        return false;
    }

    return args.isActionOwner || args.isAdmin;
}
```

Reply:
Looks much cleaner to me @cubuspl42 🚀